tendermint/src/amino_types/vote.rs

```diff
@@ -49,6 +51,30 @@ impl Vote {
     }
 }
 
+impl From<&vote::Vote> for Vote {
+    fn from(vote: &vote::Vote) -> Self {
+        Vote {
+            vote_type: vote.vote_type.to_u32(),
+            height: vote.height.value() as i64, // TODO potential overflow :-/
+            block_id: Some(BlockId {
```

Comment:
Hmm, I don't remember tbh. These amino types are quite old. Let me check if the encoding matches if remove the `Option`.

Reply:
So, it looks like one can simply add the `#[prost(message)]` for a nested struct if it is an `Option`. Otherwise, it seems a little more effort is necessary. e.g. I locally changed the field's type to `BlockId` instead of `Option<BlockId>`. Then the compiler (due to prost) will complain:

```
error[E0599]: no method named `as_ref` found for type `amino_types::block_id::BlockId` in the current scope
  --> tendermint/src/amino_types/vote.rs:22:28
   |
22 | #[derive(Clone, PartialEq, Message)]
   |                            ^^^^^^^ method not found in `amino_types::block_id::BlockId`
   | 
  ::: tendermint/src/amino_types/block_id.rs:11:1
   |
11 | pub struct BlockId {
   | ------------------ method `as_ref` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `as_ref`, perhaps you need to implement it:
           candidate #1: `std::convert::AsRef`

error[E0308]: mismatched types
  --> tendermint/src/amino_types/vote.rs:22:28
   |
22 | #[derive(Clone, PartialEq, Message)]
   |                            ^^^^^^^
   |                            |
   |                            expected struct `amino_types::block_id::BlockId`, found enum `std::option::Option`
   |                            this match expression has type `amino_types::block_id::BlockId`
   |
   = note: expected type `amino_types::block_id::BlockId`
              found type `std::option::Option<_>`

error[E0599]: no method named `get_or_insert_with` found for type `amino_types::block_id::BlockId` in the current scope
  --> tendermint/src/amino_types/vote.rs:22:28
   |
22 | #[derive(Clone, PartialEq, Message)]
   |                            ^^^^^^^ method not found in `amino_types::block_id::BlockId`
   | 
  ::: tendermint/src/amino_types/block_id.rs:11:1
   |
11 | pub struct BlockId {
   | ------------------ method `get_or_insert_with` not found for this

error[E0308]: mismatched types
  --> tendermint/src/amino_types/vote.rs:22:28
   |
22 | #[derive(Clone, PartialEq, Message)]
   |                            ^^^^^^^ expected struct `amino_types::block_id::BlockId`, found enum `std::option::Option`
   |
   = note: expected type `amino_types::block_id::BlockId`
              found type `std::option::Option<_>`

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: Could not compile `tendermint`.
´´´

Reply:
OK, it looks like this can be fixed by adding in a required attribute: https://github.com/danburkert/prost/blob/9551f2852e414df72339634087c46ce5a08e85b2/prost-derive/src/field/message.rs#L14-L56

I'm looking into that.

Reply:
Adding the `required` attribute, e.g.:
```rust
#[prost(message, required, tag = "4")]
pub block_id: BlockId,
```
makes the compiler happy 😄but breaks some encoding test-vectors 🤔 

```
thread 'amino_types::vote::tests::test_sign_bytes_compatibility' panicked at 'assertion failed: `(left == right)`
  left: `[15, 34, 0, 42, 11, 8, 128, 146, 184, 195, 152, 254, 255, 255, 255, 1]`,
 right: `[13, 42, 11, 8, 128, 146, 184, 195, 152, 254, 255, 255, 255, 1]`', tendermint/src/amino_types/vote.rs:317:9
```
Hmm, this feels somehow familiar. Will investigate further.

Reply:
OK, here is the catch:

In golang one can simply not initalize a field. Golang won't complain. And amino will treat this as sth. that can be skipped (and hence is optional in the protobuf sense). 

Now on the rust side you have to init a field (unless you make it an option). If you make it an Option, the encoding behaves like amino (None will be not encoded at all). If you have a regular field (e.g. BlockId), and add a `required` attribute, prost(-amino) will treat as `required` (in the protobuf sense) and signal, we got fiel blah of type blubb but nothing in there (for blockId this is fieldnumber 4 and hence `4 << 3 | 2 = 32` followed by a `0x00` for indicating that this was empty. 
Amino actually does the same but rolls back if it found out there is nothing (via that zero 0x00) unless told otherwise via `writeEmpty` (at least it seems to work [now](https://github.com/tendermint/go-amino/pull/287)):
https://github.com/tendermint/go-amino/blob/fbf776258498dbb3aa6637ca82c7fe5c7255fd2f/binary-encode.go#L496-L524

![Alt Text](http://giphygifs.s3.amazonaws.com/media/N8OXS6RFwnBkI/200w_d.gif)